frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+		self.insertion_overlay.remove(&idx).or_else(|| {
+			if self.deletion_overlay.contains(&idx) {
+				None
+			} else {
+				self.deletion_overlay.insert(idx);
+				SignedSubmissionsMap::<T>::try_get(idx).ok()
+			}
+		})
+	}
+
+	/// Iterate through the set of signed submissions in order of increasing score.
+	pub fn iter(&self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.iter().map(move |(_score, idx)| {
+			self.map_get(*idx).expect("SignedSubmissions must remain internally consistent")
+		})
+	}
+
+	/// Empty the set of signed submissions, returning an iterator of signed submissions in
+	/// arbitrary order
+	pub fn drain(&mut self) -> impl '_ + Iterator<Item = SignedSubmissionOf<T>> {
+		self.indices.clear();
+		SignedSubmissionNextIndex::<T>::kill();
+		let insertion_overlay = sp_std::mem::take(&mut self.insertion_overlay);
+		SignedSubmissionsMap::<T>::drain()
+			.filter(move |(k, _v)| !self.deletion_overlay.contains(k))
+			.map(|(_k, v)| v)
+			.chain(insertion_overlay.into_iter().map(|(_k, v)| v))
+	}
+
+	/// Decode the length of the signed submissions without actually reading the entire struct into
+	/// memory.
+	///
+	/// Note that if you hold an instance of `SignedSubmissions`, this function does _not_
+	/// track its current length. This only decodes what is currently stored in memory.
+	pub fn decode_len() -> Option<usize> {
+		SignedSubmissionIndices::<T>::decode_len()
+	}
+
+	/// Insert a new signed submission into the set.
+	///
+	/// Returns `(inserted, removed)`. `inserted` is true when the submission was inserted.
+	/// `removed` is the removed weakest submission, if any.
+	///
+	/// In the event that the new submission is not better than the current weakest according
+	/// to `is_score_better`, we do not change anything.
+	pub fn insert(
+		&mut self,
+		submission: SignedSubmissionOf<T>,
+	) -> (bool, Option<SignedSubmissionOf<T>>) {
+		let weakest = match self.indices.try_insert(submission.solution.score, self.next_idx) {
+			Ok(Some(prev_idx)) => {
+				// a submission of equal score was already present in the set;
+				// no point editing the actual backing map as we know that the newer solution can't
+				// be better than the old. However, we do need to put the old value back.
+				self.indices
+					.try_insert(submission.solution.score, prev_idx)
+					.expect("didn't change the map size; qed");
+				return (false, None);
+			}
+			Ok(None) => {
+				// successfully inserted into the set; no need to take out weakest member
+				None
+			}
+			Err((score, insert_idx)) => {
+				// could not insert into the set because it is full.
+				// note that we short-circuit return here in case the iteration produces `None`.
+				// If there wasn't a weakest entry to remove, then there must be a capacity of 0,
+				// which means that we can't meaningfully proceed.
+				let (weakest_score, weakest_idx) = match self.indices.iter().next() {
```

Comment:
I haven't read the rest of this code, but it seems like you are making an invariant that the size of the queue is never 0? if so, I think this will be a nice feature to have. As-in, if the size of the queue is set to zero, then the governance is basically disabling signed solutions temporarily. 

If you have this invariant, please put it in `feasibility_chekc` as it is important.

Reply:
Now reading the code, it seems like `return (false, None)` should handle everythign well.